bioio_base/embedded_metadata.py

```diff
@@ -0,0 +1,136 @@
+from typing import (
+    Any,
+    ClassVar,
+    Generic,
+    Optional,
+    Sequence,
+    Set,
+    TypeVar,
+    Union,
+    overload,
+)
+
+from .dimensions import DimensionNames
+
+# -----------------------------------------------------------------------------
+# Descriptor for metadata fields.
+# -----------------------------------------------------------------------------
+T = TypeVar("T")  # Generic type variable to enable type-safe storage of metadata
+
+
+class MetadataField(Generic[T]):
```

Comment:
What is the benefit of having this be a class?

Reply:
The idea was to maintain the label name and typing and have that be the type of each property in DmbeddedMetadata. This means that we can look at all the `MetadataField` properties in the class and enforce a list of allowed properties with this function 
```
def _compute_allowed_fields(cls: type) -> set[str]:
    allowed = set()
    for name, value in cls.__dict__.items():
        if isinstance(value, MetadataField):
            allowed.add(name)
            if value.private_name is not None:
                allowed.add(value.private_name)
    return allowed
```

Reply:
allowing for functionality like this 
```
   class MyReader(Reader):
       def _build_embedded_metadata(self) -> EmbeddedMetadata:
           manifest = super()._build_embedded_metadata()
           # Users can assign values only to the predefined properties.
           manifest.objective = "40x Magnification"
           manifest.column = "B3"
           manifest.well = "A1"
           # Attempting to add a new attribute like:
           # manifest.new_field = "Not Allowed" HERE
           # will raise an AttributeError.
           return manifest
```